### PR TITLE
[CI] fix horovod tests

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Horovod tests are failing with mysterious errors

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @carmocca @akihironitta @borda @awaelchli @justusschock @kaushikb11

### CODE REVIEW

**Comment by otaj** [comment]
For completeness sake of what broke - Horovod tests were broken due to NCCL, which from 2.12.7 broke when linked against dynamic `libcudart` - https://github.com/NVIDIA/nccl/issues/658

The solution is to forcefully downgrade to the last working version `2.11.4`, which will then be picked up by Horovod when compiling.

Added couple small improvements to how GPU tests are being run (and the environment being prepared)

File: .github/workflows/ci-pytorch-dockers.yml [hunk 1]
```diff
@@ -1,28 +1,14 @@
...
-      - "!dockers/README.md"
-      - "requirements.txt"
-      - "requirements/*.txt"
-      - "requirements/pytorch/*"
-      - "environment.yml"
-      - ".github/workflows/*docker*.yml"
-      - "setup.py"
-  schedule:
-    - cron: "0 0 * * *"  # at the end of every day
+  pull_request: {}
```
Comment (justusschock) [inline .github/workflows/ci-pytorch-dockers.yml:4]: need to revert this later on :)

File: .azure/gpu-tests.yml [hunk 2]
```diff
@@ -72,12 +72,16 @@ jobs:
...
         python -c "fname = 'requirements/pytorch/strategies.txt' ; lines = [line for line in open(fname).readlines() if 'bagua' not in line] ; open(fname, 'w').writelines(lines)"
+        TORCH_VERSION=$(python -c "import torch; print(torch.__version__.split('+')[0])")
         CUDA_VERSION_MM=$(python -c "import torch ; print(''.join(map(str, torch.version.cuda.split('.')[:2])))")
         CUDA_VERSION_BAGUA=$(python -c "print([ver for ver in [115,113,111,102] if $CUDA_VERSION_MM >= ver][0])")
+        python ./requirements/pytorch/adjust-versions.py requirements/pytorch/base.txt ${PYTORCH_VERSION}
+        python ./requirements/pytorch/adjust-versions.py requirements/pytorch/extra.txt ${PYTORCH_VERSION}
+        python ./requirements/pytorch/adjust-versions.py requirements/pytorch/examples.txt ${PYTORCH_VERSION}
         pip install "bagua-cuda$CUDA_VERSION_BAGUA>=0.9.0"
-        pip install -e .[strategies]
+        pip install -e .[strategies] --find-links https://download.pytorch.org/whl/cu${CUDA_VERSION_MM}/torch_stable.html
```
Comment (justusschock) [inline .azure/gpu-tests.yml:82]: very nice!

File: .github/workflows/ci-pytorch-dockers.yml [hunk 3]
```diff
@@ -1,28 +1,14 @@
...
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}-${{ github.head_ref }}-${{ github.event_name }}
   cancel-in-progress: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
 
 env:
-  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
+  PUSH_TO_HUB: true
```
Comment (justusschock) [inline .github/workflows/ci-pytorch-dockers.yml:11]: and this as well :)

File: dockers/base-cuda/Dockerfile [hunk 4]
```diff
@@ -50,8 +54,10 @@ RUN \
         libopenmpi-dev \
         openmpi-bin \
         ssh \
-    && \
+        libnccl2=$TO_INSTALL_NCCL \
+        libnccl-dev=$TO_INSTALL_NCCL
 
+RUN \
```
Comment (justusschock) [inline dockers/base-cuda/Dockerfile:60]: why have this separately? This would just produce more layers -> more overhead

**Comment by otaj** [comment]
This will be ready to merge once Docker images are built and pushed to dockerhub

File: dockers/base-cuda/Dockerfile [hunk 5]
```diff
@@ -61,8 +67,9 @@ RUN \
     && \
 
     update-alternatives --install /usr/bin/python${PYTHON_VERSION%%.*} python${PYTHON_VERSION%%.*} /usr/bin/python${PYTHON_VERSION} 1 && \
-    update-alternatives --install /usr/bin/python python /usr/bin/python${PYTHON_VERSION} 1 && \
+    update-alternatives --install /usr/bin/python python /usr/bin/python${PYTHON_VERSION} 1
 
+RUN \
```
Comment (justusschock) [inline dockers/base-cuda/Dockerfile:72]: same here

**Review by justusschock — APPROVED** [review]
LGTM

File: .github/workflows/ci-pytorch-dockers.yml (hunk 1 above)
Comment (otaj) [reply .github/workflows/ci-pytorch-dockers.yml:4]: I know, ready for it to be last commit :)

File: dockers/base-cuda/Dockerfile (hunk 4 above)
Comment (otaj) [reply dockers/base-cuda/Dockerfile:60]: Docker was complaining to me that in future releases, empty continuation lines are going to be errors in dockerfiles. Yes, it is a bit of overhead, but not too much

File: dockers/base-cuda/Dockerfile (hunk 5 above)
Comment (otaj) [reply dockers/base-cuda/Dockerfile:72]: Same as previous comment

File: dockers/base-cuda/Dockerfile (hunk 4 above)
Comment (akihironitta) [reply dockers/base-cuda/Dockerfile:60]: I would remove the empty lines rather than splitting it into multiple `RUN`s 👀

File: .azure/gpu-tests.yml [hunk 6]
```diff
@@ -72,12 +72,16 @@ jobs:
...
         pip install -U deepspeed  # TODO: remove when docker images are upgraded
```
Comment (carmocca) [inline .azure/gpu-tests.yml:83]: May I use the opportunity to remove this TODO?

```suggestion
```

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: dockers/base-cuda/Dockerfile (hunk 4 above)
Comment (otaj) [reply dockers/base-cuda/Dockerfile]: Alright, removed empty lines

File: .azure/gpu-tests.yml (hunk 6 above)
Comment (otaj) [reply .azure/gpu-tests.yml]: Applied the suggestion

**Review by awaelchli — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by Borda — CHANGES_REQUESTED** [review]
Are we aware that the GOP tests are skipped?
https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=92435&view=logs&jobId=3f274fac-2e11-54ca-487e-194c91f3ae9f

**Review by Borda — APPROVED** [review]
